### PR TITLE
DAOS-11404 vos/dtx: Introduce local transactions

### PR DESCRIPTION
**Pull request by janekmi**
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

- - -
This change is [Reviewable](https://reviewable.io/reviews/janekmi/daos/23)

### CODE REVIEW

File: src/dtx/SConscript [hunk 1]
```diff
@@ -13,7 +13,6 @@ def scons():
     denv = env.Clone()
 
     denv.require('argobots', 'protobufc')
-    denv.require('pmdk', headers_only=True)
```
Comment (grom72) [inline src/dtx/SConscript:16]: Is this a part of this PR?

File: src/dtx/dtx_common.c [hunk 2]
```diff
@@ -861,10 +862,15 @@ dtx_handle_init(struct dtx_id *dti, daos_handle_t coh, struct dtx_epoch *epoch,
 
 	dtx_shares_init(dth);
 
-	dth->dth_xid = *dti;
+	if (!local) {
+		dth->dth_xid        = *dti;
```
Comment (grom72) [inline src/dtx/dtx_common.c:866]: We use fth->dth_xid also for local case - see line 869

**Review by grom72 — CHANGES_REQUESTED** [review]

**Review by grom72 — CHANGES_REQUESTED** [review]
> Reviewed 6 of 20 files at r1, 2 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: 8 of 20 files reviewed, 8 unresolved discussions (waiting on @janekmi)

___
*[`src/dtx/dtx_common.c` line 869 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMjHrn-P5P8tbv-sG3:-NlMjHrn-P5P8tbv-sG4:b-5pkbx) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L869)):*
> ```C
> 		dth->dth_leader_oid = *leader_oid;
> 	} else {
> 		dth->dth_xid.dti_hlc = 1;
> ```

Dont we need to assign first:
`dth->dth_xid        = *dti;`?



_Code quote:_
```C
		dth->dth_xid.dti_hlc = 1;
```
___
*[`src/dtx/dtx_common.c` line 938 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMkE50DcYKmbcHHZg6:-NlMkE51BP4SAnq7gpFF:b4hazcb) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L938)):*
> ```C
> 		rc = vos_dtx_local_begin(dth, coh);
> 		if (rc) {
> 			goto error;
> ```

Shall we provide an ERROR message?



_Code quote:_
```C
			goto error;
```
___
*[`src/dtx/dtx_common.c` line 1523 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMkshA1FQ6FJ3uTxN0:-NlMkshA1FQ6FJ3uTxN1:bno9ntc) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L1523)):*
> ```C
> 
> 	if (dth->dth_local)
> 		dth->dth_local_complete = 1;
> ```

Why it is not a part of `vos_dtx_local_end`?



_Code quote:_
```C
	if (dth->dth_local)
		dth->dth_local_complete = 1;
```
___
*[`src/dtx/SConscript` line 16 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMl1_h3vvoMTfTkeuK:-NlMl1_h3vvoMTfTkeuL:btssgq) ([raw file](https://github.com/janekmi/daos/blob/2207e336971a2f0832bade728041afe50c17e9d2/src/dtx/SConscript#L16)):*
> ```Python
> 
>     denv.require('argobots', 'protobufc')
>     denv.require('pmdk', headers_only=True)
> ```

Is it a part of this PR?



_Code quote:_
```Python
   denv.require('pmdk', headers_only=True)
```
___
*[`src/dtx/tests/dts_local.c` line 159 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMmvCtEPgzOTHBEmLy:-NlMmvCtEPgzOTHBEmLz:bq1opmr) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/tests/dts_local.c#L159)):*
> ```C
> 			/** On abort,
> 			 * the value is not meant to be found hence the output size == 0 */
> 			assert_int_equal(iod.iod_size, 0);
> ```

I remember (not 100% sure) that there was a problem that `iod_size==0` but there is some data in buffer.



_Code quote:_
```C
			/** On abort,
			 * the value is not meant to be found hence the output size == 0 */
			assert_int_equal(iod.iod_size, 0);
```
___
*[`src/dtx/tests/dts_local.c` line 208 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMo2mL8pbJJVdDpafS:-NlMo2mL8pbJJVdDpafT:b-dzrd71) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/tests/dts_local.c#L208)):*
> ```C
> 		buf_size = 100;
> 
> 	rc = cmocka_run_group_tests_name(test_name, local_tests_all, setup_io, teardown_io);
> ```

would it be better to have `setup_dtx_local` which calls `setup_io` instead of long setup sequence in  `local_transaction`?
in such case, we could have a separate test for normal and for abort case



_Code quote:_
```C
cmocka_run_group_tests_name
```

**Review by janekmi — COMMENTED** [review]
> Reviewed 11 of 20 files at r1, 8 of 9 files at r2, 8 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: 19 of 20 files reviewed, 8 unresolved discussions (waiting on @grom72)

___
*[`src/dtx/dtx_common.c` line 866 at r2](https://reviewable.io/reviews/janekmi/daos/23#gh-1422307483:-NlNvMstAwFSRZd--aJ4:b-z4gjaf) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L866)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

We use fth->dth_xid also for local case - see line 869
</blockquote></details>

Yes, we do. However, the local variation doesn't have a transaction ID generated on the client side. The `dti` pointer in the local case is NULL. So, we can not use it to initialize `dth->dth->xid`.



_Code snippet:_
```C
/**
 * DAOS two-phase commit transaction identifier,
 * generated by client, globally unique.
 */
struct dtx_id;
```
___
*[`src/dtx/dtx_common.c` line 869 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMjHrn-P5P8tbv-sG3:-NlNw5RsDvt7uvr8qZ4O:bv4iose) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L869)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Dont we need to assign first:
`dth->dth_xid        = *dti;`?
</blockquote></details>

Note the `dth->dth_xid` is a value, not a pointer. The assignment above is just to copy both fields at the same time. But, since we do not have the user-provided `dti` in this case the local variation has to do the minimum to slip through the various DTX checks. Other than that the `dti_hlc` value is meaningless in the case of the local transactions.

The HLC concept makes sense only in the distributed context: https://sergeiturukin.com/2017/06/26/hybrid-logical-clocks.html
___
*[`src/dtx/dtx_common.c` line 938 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMkE50DcYKmbcHHZg6:-NlNxCEf1pf6J_ueQHwb:b-bsq02s) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L938)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Shall we provide an ERROR message?
</blockquote></details>

I believe `vos_dtx_local_begin()` does the thing. If you think a particular error case is not covered please point it out directly.
___
*[`src/dtx/SConscript` line 0 at r2](https://reviewable.io/reviews/janekmi/daos/23#gh-1422306157:-NlNzZeq56mMRUt7tnXj:bdazlwk) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/SConscript#L0)):*
Can you repost this comment?

> (Reviewable was unable to map this GitHub inline comment thread to the right spot — sorry!)
___
*[`src/dtx/SConscript` line 16 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMl1_h3vvoMTfTkeuK:-NlNzzsi9hxF2REcdq2c:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/2207e336971a2f0832bade728041afe50c17e9d2/src/dtx/SConscript#L16)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Is it a part of this PR?
</blockquote></details>

Done.
___
*[`src/vos/vos_aggregate.c` line 1428 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlO4il7Akml5Pqfi5Uo:-NlO4il7Akml5Pqfi5Up:b-3akilf) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/vos/vos_aggregate.c#L1428)):*
> ```C
> 
> 	/* Publish SCM reservations */
> 	rc = vos_publish_scm(vos_obj2umm(obj), io->ic_rsrvd_scm, true);
> ```

@grom72 you asked offline whether this bit is also related to this PR. And the answer is: yes. Since local transactions are not limited to a single container but rather to a pool and `vos_publish_scm()` doesn't care about the container it needs the UMM targeting the underlying medium, the provided argument had to be altered accordingly.

You can see more in the `vos_io.c` file.



_Code quote:_
```C
rc = vos_publish_scm(vos_obj2umm(obj), io->ic_rsrvd_scm, true);
```

**Review by janekmi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: 19 of 20 files reviewed, 8 unresolved discussions (waiting on @grom72)

___
*[`src/dtx/dtx_common.c` line 1523 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMkshA1FQ6FJ3uTxN0:-NlOblC86QmUYmReX8aa:b-gtzitl) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/dtx_common.c#L1523)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Why it is not a part of `vos_dtx_local_end`?
</blockquote></details>

Sorry to say that but it is pretty convoluted.

1. `vos_dtx_local_end()` is only an accessory for `dtx_end()` to finalize the transaction if the transaction concludes without errors.
2. Otherwise, `dtx_end()` calls `vos_dtx_cleanup()` which calls `vos_tx_end()` directly. So, there is no `vos_dtx_local_end()` on the way to set this flag.

In general, I am open to discussing the bits' placement since I am pretty unsatisfied with the overall look and feel of it. But I am afraid it is the best that we can have right now.

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 8 unresolved discussions (waiting on @grom72)

**Review by janekmi — COMMENTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 8 unresolved discussions (waiting on @grom72)

___
*[`src/dtx/tests/dts_local.c` line 159 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMmvCtEPgzOTHBEmLy:-NlTyBnRD26vuuLhwJ8d:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/tests/dts_local.c#L159)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

I remember (not 100% sure) that there was a problem that `iod_size==0` but there is some data in buffer.
</blockquote></details>

Done.
___
*[`src/dtx/tests/dts_local.c` line 208 at r2](https://reviewable.io/reviews/janekmi/daos/23#-NlMo2mL8pbJJVdDpafS:-NlTyGXs385QSFXhBHCG:b6xbh5v) ([raw file](https://github.com/janekmi/daos/blob/0a9928c4d6694c6cbc3058964f40fde65b08230a/src/dtx/tests/dts_local.c#L208)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

would it be better to have `setup_dtx_local` which calls `setup_io` instead of long setup sequence in  `local_transaction`?
in such case, we could have a separate test for normal and for abort case
</blockquote></details>

I am not convinced it would be better to have separate normal and abort cases. I believe it is greatly simplified without introducing this split. Both cases are mostly the same.

**Review by grom72 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 9 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 2 unresolved discussions (waiting on @janekmi)

___
*[`src/dtx/tests/dts_local.c` line 117 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlUorvu08aWuOw7OgXH:-NlUorvu08aWuOw7OgXI:b45h) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/tests/dts_local.c#L117)):*
> ```C
> 	memset(buf, 'x', BUF_SIZE - 1);
> 	buf[BUF_SIZE - 1] = '\0';
> }
> ```





_Suggestion:_
```C
/**
 * Invalidate contents of the provided buffer.
 *
 * Note: The buffer filled with 'x'-es is considered invalidated.
 */
static void
invalidate_buf(char *buf, size_t size)
{
	memset(buf, 'x', size - 1);
	buf += size - 1;
	*buf = '\0';
}
```
___
*[`src/dtx/tests/dts_local.c` line 123 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlUpyBs0xDfV10B3YYi:-NlUpyBs0xDfV10B3YYj:undefined) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/tests/dts_local.c#L123)):*
> ```C
>       d_sg_list_t *sgl)
> {
> 	invalidate_buf(sgl->sg_iovs[0].iov_buf);
> ```





_Suggestion:_
```C
invalidate_buf(sgl->sg_iovs[0].iov_buf, BUF_SIZE);
```

**Review by osalyk — COMMENTED** [review]
> Reviewed 11 of 20 files at r1, 2 of 9 files at r2, 7 of 9 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 4 unresolved discussions (waiting on @janekmi)

___
*[`src/dtx/dtx_common.c` line 1161 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlXIA3W5DGMDHYc4NxI:-NlXIA3W5DGMDHYc4NxJ:undefined) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/dtx_common.c#L1161)):*
> ```C
> 
> 	dth = &dlh->dlh_handle;
> 	rc  = dtx_handle_init(
> ```





_Suggestion:_
```C
rc = dtx_handle_init(
```
___
*[`src/dtx/tests/dts_local.c` line 51 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlXJxcOAgdVV6t-S5sZ:-NlXJxcP8pKZiGPxYfaj:b59q62t) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/tests/dts_local.c#L51)):*
> ```C
> 	struct dts_local_args *la       = &local_args;
> 	int                    int_flag = is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64);
> 	int                    rc;
> ```

I have a feeling that something is wrong with the formatting



_Code quote:_
```C
	struct io_test_args *arg = *state;
	struct dts_local_args *la       = &local_args;
	int                    int_flag = is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64);
	int                    rc;
```

**Review by grom72 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 5 unresolved discussions (waiting on @janekmi)

___
*[`src/dtx/tests/dts_local.c` line 300 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlXjxmf-Vfb-H2QVZvg:-NlXjxmg8wWLhOLFFVhI:undefined) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/tests/dts_local.c#L300)):*
> ```C
> 
> 	/** initialize the global invalid buffer */
> 	invalidate_buf(invalid_data);
> ```





_Suggestion:_
```C
invalidate_buf(invalid_data, BUF_SIZE);
```

**Review by janekmi — COMMENTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23)* status: all files reviewed, 2 unresolved discussions (waiting on @osalyk)

___
*[`src/dtx/dtx_common.c` line 1161 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlXIA3W5DGMDHYc4NxI:-NlY0_Yx3xWKpbZVv_jF:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/dtx_common.c#L1161)):*
> ```C
> 
> 	dth = &dlh->dlh_handle;
> 	rc  = dtx_handle_init(
> ```

Done.
___
*[`src/dtx/tests/dts_local.c` line 51 at r4](https://reviewable.io/reviews/janekmi/daos/23#-NlXJxcOAgdVV6t-S5sZ:-NlXzSjI6qW30uSvVnmK:beyjv1q) ([raw file](https://github.com/janekmi/daos/blob/acfc24ea6c1a4e30d4393c54ff2a68c4d0311b78/src/dtx/tests/dts_local.c#L51)):*
<details><summary><i>Previously, osalyk (Oksana Sałyk) wrote…</i></summary><blockquote>

I have a feeling that something is wrong with the formatting
</blockquote></details>

TBH I feel the same. But you can't argue with `clang-format`. Sorry.

**Review by grom72 — COMMENTED** [review]
:goat:

> Reviewed 3 of 20 files at r1, 1 of 9 files at r2, 4 of 9 files at r3, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/23#-:-NlY4i1r6OIGMRB4wSyX:bnd2vnu)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @janekmi)

**Pull request closed without merging**